article/2020/how-to-write-clean-code-lessons-learnt-from-the-clean-code-robert-c-martin.md

```diff
@@ -165,31 +164,31 @@ public class Circle implements Shape {
 }
 ```
 
-Now we can easily add new Shapes i.e. data structures as compared to previous case. And if we have to add perimeter() function in only one Shape, we are forced to implement that function in all the Shapes as Shape class is an interface containing area() and perimeter() function. This means:
+现在，与之前的案例相比，我们能很轻松的添加新形状，即数据结构。而且，如果我们只需在一个 Shape 类中添加 perimeter() 方法，则我们就必须在所有 Shapes 类中实现该函数，因为 Shapes 类是一个包含 area() 和 perimeter() 方法的接口。这意味着：
 
-> **Data structures makes it easy to add new functions without changing the existing data structures. OO code(using objects), makes it easy to add new classes without changing existing functions.**
+> **数据结构便于在不改动既有数据结构的前提下添加新函数。面向对象代码（使用对象）便于在不改动既有函数的前提下添加新类。**
 
-The complimentary is also true:
+反过来讲也说得通：
 
-> **Procedural code(using data structures) makes it hard to add new data structures because all the functions must change. OO code makes it hard to add new functions because all the classes must change.**
+> **过程式代码（使用数据结构的代码）难以添加新的数据结构，因为必须修改所有的函数，面向对象代码难以添加新函数，因为必须修改所有的类。**
 
-So, the things that are hard for OO are easy for procedures, and the things that are hard for procedures are easy for OO!
+因此，对于面向对象困难的事情对于过程来说很容易，对于面向对象容易的事情对于过程来说很困难。
 
-In any complex system there are going to be times when we want to add new data types rather than new functions. For these cases objects and OO are most appropriate. On the other hand, there will also be times when we’ll want to add new functions as opposed to data types. In that case procedural code and data structures will be more appropriate.
+在任何复杂的系统中，我们有时会希望能够灵活的添加新的数据类型而不是新的函数。在这种情况下，对象和面向对象就是最合适的。另外一些时候，我们希望能灵活的添加新函数而不是数据类型。在这种情况下，过程式代码和数据结构将会更加合适。
```

Comment:
```suggestion
在任何复杂的系统中，比起添加新函数来说我们有时会更希望能够灵活的添加新的数据类型。在这种情况下，对象和面向对象就是最合适的。另外一些时候，我们希望能灵活的添加新函数而不是数据类型。在这种情况下，面向过程的代码和数据结构将会更加合适。
```

Reply:
《代码整洁之道》和本文作者的英文表达有一些差异，但是表达的意思是一样的。
![image](https://user-images.githubusercontent.com/18589232/81488851-384a3a00-92a1-11ea-9431-bba64944e70f.png)

Reply:
这是《clean code》的原文：
In any complex system there are going to be times when we want to add new data types rather than new functions. For these cases objects and OO are most appropriate. On the other hand, there will also be times when we’ll want to add new functions as opposed to data types. In that case procedural code and data structures will be more appropriate.